File: Python/uploadFiles.py
'''
DISCLAIMER OF WARRANTIES.
 This code is sample code created by IBM Corporation. IBM grants you a
 nonexclusive copyright license to use this sample code example. This
 sample code is not part of any standard IBM product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. This example has not been thoroughly tested under all
 conditions. IBM, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. The code is provided "AS IS",
 without warranty of any kind. IBM shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if IBM
 has been advised of the possibility of such damages. If you do not agree with
 these terms, do not use the sample code.

 Copyright IBM Corp. 2019 All Rights Reserved.

 To run, see README.md
'''
import os, json, copy, re

import requests, time
import datetime as dt
from base64 import b64encode
from readConfigJSON import readJSON
from loggingHandler import logger
from ssl import SSLError


'''
    Function to loop through input directory and upload all the files into the Content Analyzer API environment
'''
def uploadFiles():
    configuration, configuration_settings = readJSON()
    if (configuration):
        starttime = dt.datetime.now()
        dir_path = configuration_settings["directory_path"]
        count = 0

        errors = []
        upload_url = configuration_settings['main_url']
        accepted_extensions = ["pdf", "jpeg", "jpg", "png", "pneg", "tiff", "tif", "docx", "ppt", "doc", "pptx", "xls", "xlsx"]
        file_types =  configuration_settings['file_type'] if 'file_type' in configuration_settings and type(configuration_settings['file_type'] is list) and len(configuration_settings['file_type']) > 0 else accepted_extensions
        output_results = []
        for subdir, dirs, files in os.walk(dir_path):
            for file in files:
                file_path = os.path.join(subdir, file)
                new_file = copy.copy(file)

                file_split = new_file.rsplit(".")
                file_extension = str(file_split[-1].strip().lower())
                old_file_name = new_file.replace("." + file_extension, '').strip()
                file_name = re.sub('[^A-Za-z0-9 _]+', ' ', old_file_name) + "." + str(file_extension)
                new_file_path = os.path.join(subdir, file_name)
                if(file_extension != "" and file_extension in accepted_extensions and file_extension in file_types):
                    count += 1
                    try:
                        logger.info("Uploading {0} ".format(new_file_path))
                    except:
                        pass

                    files = {'file': (file_name, open(file_path, 'rb'), "multipart/form-data")}
                    headers = {
                        'apiKey': configuration_settings['api_key']
                    }

                    if "function_id" in configuration_settings:
                            credentials = ('%s:%s' % (
                            configuration_settings["function_id"], configuration_settings["password"]))
                            encoded_credentials = b64encode(credentials.encode('ascii'))
                            headers['Authorization'] = 'Basic %s' % encoded_credentials.decode("utf-8")

                    dict_object = {"filename": file_name, "path": os.path.basename(subdir), "full_path": os.path.join(os.path.abspath(subdir), old_file_name + "." + str(file_extension))}

                    # Make request
                    try:
                        response = requests.request("POST", upload_url, files=files,
                                data={'jsonOptions': configuration_settings['json_options'], 'responseType': configuration_settings['output_options']}, headers=headers, verify=configuration_settings['ssl_verification'])
                        if response.status_code >= 400:
                            logger.error("HTTP error {0} occurred when uploading file: {1} ".format(str(response.status_code), file_path))
                            dict_object.update({"error": response.text})
                            errors.append(dict_object)
                        else:
                            dict_object.update({"response": response.text, "output_type": configuration_settings['output_options'].split(",")})
                            output_results.append(dict_object)

                    except SSLError as sslerror:
                        logger.error("SSL error was thrown due to certificate failure, set ssl_verification to false in configuration config.json file.")
                        dict_object.update({"error": str(sslerror)})
                        errors.append(dict_object)
                    except Exception as ex:
                        dict_object.update({"error": str(ex)})
                        errors.append(dict_object)
                        logger.error("An error occurred when trying to upload file " + file_path)
                        logger.debug(ex, exc_info=True)
                        pass

                endtime = dt.datetime.now()
                seconds = (endtime - starttime).total_seconds()
                result = {"starttime": str(starttime), "endtime": str(endtime), "no_of_files": count,
                          "output_results": output_results, "no_output_results": len(output_results),
                          "total_upload_seconds": seconds, "upload_errors": errors, "no_of_upload_errors": len(errors)}
                json.dump(result, open(os.path.join(os.getcwd(), "output.json"), 'w'), indent=4)
        logger.info("Upload status reported in output.json")
        if count == 0:
            logger.error("No files found in directory_path, " + dir_path)
            return False
        elif len(errors) > 0:
            logger.error("Errors occurred during upload.")
            return True
        else:
            logger.info("Done uploading {0} files".format(count))
            return True
    else:
        logger.error("Check your configuration file (config.json) for correct format and valid parameters")
        return False


if __name__ == '__main__':
    logger.info("Logs can be found in current directory (processing.log)")
    try:
        uploadFiles()
    except Exception as ex:
        logger.error("An error occurred, please check logs for more details")
        logger.debug(ex, exc_info=True)

File: ContentAnalyzer/C#/ContentAnalyzer/UploadFiles.cs
/*disclaimer of warranties.
 this code is sample code created by ibm corporation. ibm grants you a
 nonexclusive copyright license to use this sample code example. this
 sample code is not part of any standard ibm product and is provided to you
 solely for the purpose of assisting you in the development of your
 applications. this example has not been thoroughly tested under all
 conditions. ibm, therefore cannot guarantee nor may you imply reliability,
 serviceability, or function of these programs. the code is provided "as is",
 without warranty of any kind. ibm shall not be liable for any damages
 arising out of your or any other parties use of the sample code, even if ibm
 has been advised of the possibility of such damages. if you do not agree with
 these terms, do not use the sample code.
 copyright ibm corp. 2019 all rights reserved.
 to run, see readme.md
 * */

using System;
using System.Linq;
using System.Net;
using System.IO;
using RestSharp;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NLog;

namespace ContentAnalyzer
{
    public class UploadFiles
    {
        private static Logger logger = LogManager.GetCurrentClassLogger();

        /// <summary>
        ///Input params in JSON configuration file (example):
        ///</summary>       
        public class Config
        {
            public string functionalID { get; set; }
            public string password { get; set; }
            public string LDAP_userName { get; set; }
            public string LDAP_password { get; set; }
            public string apiKey { get; set; }
            public string mainURL { get; set; }
            public string responseType { get; set; }
            public string jsonOptions { get; set; }
            public string directoryPath { get; set; }
            public string outputDirectoryPath { get; set; }

        }
        public static Config configParams;

        /// <summary>
        /// Read the configuration.json file to get the request parameters required for the uploading request.
        /// start uploading the files, return the response and save the response into the output.json file. 
        /// </summary>
        /// <param>configFilePath</param>
        ///<returns>analyzerId in the output.json, which will be used in the get request</returns>
        public Boolean UploadFile(string configFilePath)
        {
            var startTime = DateTime.Now;

            ServicePointManager.ServerCertificateValidationCallback += (sender, certificate, chain, sslPolicyErrors) => true;
            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;

            DirectoryInfo currentDir = new DirectoryInfo(Directory.GetCurrentDirectory());
            string configFile = currentDir.Parent.Parent.FullName + @"/NLog.config";

            string result = String.Empty;
            string outputDir = @"../../../output.json";
            configParams = new Config();
            string jsonParams = File.ReadAllText(@configFilePath);
            configParams = (JsonConvert.DeserializeObject<Config>(jsonParams));

            string functionalID = configParams.functionalID;
            string password = configParams.password;
            string LDAP_userName = configParams.LDAP_password;
            string LDAP_password = configParams.LDAP_password;
            string mainURL = configParams.mainURL.Trim();
            string apiKey = configParams.apiKey;
            string responseType = configParams.responseType;
            string jsonOptions = configParams.jsonOptions;
            string directoryPath = configParams.directoryPath;

            string outputDirectoryPath = @configParams.outputDirectoryPath;
            JArray output = new JArray();
            JArray errors = new JArray();

            string url = mainURL + "/contentAnalyzer";
            string auth = "";
            if (functionalID != "" && password != "") {
               auth = Base64Encode(functionalID + ":" + password);
            }
            else if (LDAP_userName != "" && LDAP_password != "")
            {
                auth = Base64Encode(LDAP_userName + ":" + LDAP_password);
            } 
            if (Directory.Exists(directoryPath))
            {
                string[] extensions = new[] { ".jpg", ".jpeg", ".tif", ".tiff", ".png", ".pdf", ".doc", ".docx" };
                DirectoryInfo di = new DirectoryInfo(@directoryPath);
                FileInfo[] files = di.GetFiles().Where(f => extensions.Contains(f.Extension.ToLower())).ToArray();
                int success_count = 0;
                int failed_count = 0;           
                foreach (FileInfo file in files)
                {             
                    JObject dictObject = new JObject();
                    dictObject["filename"] = file.Name;

                    var client = new RestClient(url);
                    var request = new RestRequest(Method.POST);

                    request.AddHeader("Authorization", "Basic " + auth);
                    request.AddHeader("apiKey", apiKey);
                    request.AddHeader("content-type", "multipart/form-data");
                    request.AddParameter("responseType", responseType);
                    request.AddParameter("jsonOptions", jsonOptions);
                    string filepath = file.DirectoryName + "/" + file.Name;
                    //bool fileExist = (File.Exists(filepath));
                    request.AddFile("file", filepath);
                    IRestResponse response = client.Execute(request);

                    if (response.StatusCode == System.Net.HttpStatusCode.Accepted)
                    { 
                        success_count += 1;
                        dictObject["upload_response"] = JObject.Parse(response.Content);
                        output.Add(dictObject);

                    } else {
                        failed_count += 1;
                        logger.Error("Http error occurred when uploading file " + file);
                        dictObject["upload_error"] = response.Content;
                        errors.Add(dictObject);

                    }
                }

                var endTime = DateTime.Now;
                JObject resultUpload = new JObject();
                resultUpload["startTime"] = startTime;
                resultUpload["no_files"] = success_count + failed_count;
                resultUpload["upload_results"] = output;
                resultUpload["total_upload_seconds"] = (endTime - startTime).TotalSeconds;
                resultUpload["upload_errors"] = errors;

                System.IO.File.WriteAllText(@outputDir, JsonConvert.SerializeObject(resultUpload));
                if (success_count == 0)
                {
                    logger.Error("No file uploaded successfully.");
                    return false;
                }
                else
                {
                    logger.Info("Uploaded all the files");
                    return true;
                }
                
            }
            logger.Error("No file exist in the directory");
            return false;
        }

        // To convert the basic auth
        public static string Base64Encode(string plainText)
        {
            var plainTextBytes = System.Text.Encoding.UTF8.GetBytes(plainText);
            return System.Convert.ToBase64String(plainTextBytes);
        }
    }
}
